active/0024-message-validation.md

```diff
@@ -0,0 +1,132 @@
+# Message Filter and Validation
+
+## Changelog
+
+* 2023-10-31: @zmstone Initial draft message validation, filter and transformation.
+* 2022-12-19: @zmstone Limit the scope to filter and validation.
+
+## Abstract
+
+A new feature for EMQX v5 to validate messages.
+
+## Motivation
+
+### Message Validation
+
+Technically, by using EMQX rules engine, it is possible to validate the incoming
+MQTT messages before sending the message off to data bridges.
+
+For instance, one can write a rules engine SQL like below to
+drop messages which are not avro encoded and also when the 'val' field is not greater than 10.
+
+```
+SELECT schema_decode(my_avro_schema, payload) as decoded from t/# where decoded.val > 10
+```
+
+However, currently rules are only used in data bridges, but they do not interfere MQTT message pub/sub.
+For example, if there is client subscribed to `t/#` it would still receive the avro encoded message.
+
+If we also want to stop the message being published, then there would be a need for a message republish action,
+and the publisher and subscriber will likely be forced to use different topics.
+e.g. publisher publishes to `t1/...`, republish action send it to `t2/...` for subscribers to receive.
+
+This can work, but not easy to use.
+
+### Message filter
+
+Message filtering can be achieved by confugring message validation rule to drop invalid messages.
+
+## Design
+
+This data validation step should happen after authorization and before rules engine.
+Probably can make use of the 'message.publish' with a hook priority which
+positions the callback after authorization, but before data bridges.
+
+### High level design
+
+* Make use of the rules engine SQL, but without the `FROM` clause as it's always the MQTT message being th input.
+
+Each data validation rule should have a unique ID. And each rule should have a input topic-filter name which
+can be used to build a topic index for quick look-up.
+
+The common parts can be described as hocon config below:
+
+### Message Validation
+
+```
+validation1 {
```

Comment:
Would it be better to use an array? When topics overlap, they can be prioritized according to the order of the array
```
message_validation = [
  { name = "", "topics" = ["t/#"] ... },
  { name = "", "topics" = ["t/1"] ... },
]
```

Reply:
the plan, is to index topics using the trie index.

if we want to preserve order, there would be a performance issue if for example when a thousand validations defined, but only one or two would match for each message.

Reply:
I think it's very similar to Rules, where through array configuration, a clear execution order can be provided to the users. 
The configuration still uses an array, and after saving, it's converted into a trie for indexing. The trie is used to find all matching rules, but the execution of the rules is still done according to the order in the array.
Is this feasible?